Add the initial draft for inference-using-bounds (#4140)
dart-lang/language: accepted/future-releases/3009-inference-using-bounds/feature-specification.md

> # Inference using bounds

Author: Thanks for all the comments and for pointing that the update should also be made to the grounded solution for a set of type variables. I've updated the text accordingly.

Author: Chloe Stefantsova

Status: Accepted

> Version 0.1 (see [CHANGELOG](#CHANGELOG) at end)

Reviewer: Personally I would start at 1.0. To me, "0.1" means it's a draft, and considering that we've already pretty much committed to doing this feature, it seems like we're well out of "draft" territory 😃

Author: Thanks for the suggestion. I've updated the file accordingly.

Experiment flag: inference-using-bounds

##  Motivating example

> The motivating example is discovered and discussed in [#3009][] (Type inference does not solve some constraints involving F-bounds). Consider the following code:

Author: Thanks for the suggestion, Erik! I wasn't sure which style to follow and defaulted to no line breaks. I've updated the .md document as you suggested.

Both the constraint set `{C <# X}` and the empty preliminary solution from the downwards phase `{X = _}` are passed to the algorithm described in [Constraint solution for a set of type variables][CSSTV]. The algorithm starts by checking whether the preliminary solution is known (that is, it doesn't contain the unknown type `_`), and the check fails. Then the algorithm solves the constraints for `X`, yielding `X = C` as the solution. Since `C` doesn't contain `_`, the algorithm checks if `X` has a bound (which it does: the bound of `X` is `A<X>`), computes the best-effort approximation of that bound (which is `A<_>`), adds the bound to the existing set of constraints for `X` (resulting in `{C <# X, X <# A<_>}`), and tries to solve the updated constraint set for `X` yielding the same result `X = C`.

Note that adding the best-effort approximation of the bound and solving the constraint set once again didn't change the outcome. ***In general, lower bounds are preferred by the type inference, and the best-effort approximation is added as an upper-bound constraint.***

> It is notable that even though the example program contains all the type information needed to infer `B` as the type parameter, the front ends following the specification infers `C`, resulting in a compile-time error during the type checks of the type arguments against the bounds.

Author: Thanks! Fixed.

The picture below shows the details of the type inference process for the motivating example using the schematic representation.

```mermaid
block-beta
  columns 8
  space downwards["Downwards phase"]:3 upwards["Upwards phase"]:3 space

  space:4
  newConstraint["X <: A<_>"]
  space:3

  overallInput["Input"]
  space
  csstvDownwards["CSSTV"]
  space:2
  csstvUpwards["CSSTV"]
  space
  overallOutput["Output"]

  space:8

  space:4
  constraintGeneration["X <: A<_>"]
  space:3

  space:2
  boundsDownwards["X extends A<X>"]
  space:2
  boundsUpwards["X extends A <X>"]
  space:2

  downwards -- "∅" --> csstvDownwards
  upwards -- "C <: X" --> csstvUpwards

  overallInput -- "{X = _}" --> csstvDownwards
  csstvDownwards -- "{X = _}" --> csstvUpwards
  csstvUpwards -- "{X = C}" --> overallOutput

  constraintGeneration --> newConstraint
  boundsUpwards --> constraintGeneration

  boundsDownwards --> csstvDownwards
  boundsUpwards --> csstvUpwards
```

> 

Author: Good idea, Erik! I've added a clarifying paragraph and some labels to the graph, to make the intention clear. Hopefully, it reads better now.

## Proposed changes

We propose generating more constraints from existing constraints and the bound. The original example is solved by this proposal and a few more beneficial examples are discussed afterwards.

> In the motivating example, adding the best-effort bound approximation didn't affect the result since the set of constraints for `X` already contained a lower-bound constraint for `X` (`C <# X`), and that constraint had the priority over the upper-bound "best-effort" approximation constraint, since type `C` is fully known (i.e. doesn't contain `_`). In the current proposal ***we aim to derive a lower-bound constraint by combining the already existing lower-bound constraint with the actual bound***.

Reviewer 2: These sentences could be made a bit shorter, which would probably help the reader:
```suggestion
In the motivating example, adding the best-effort bound approximation didn't affect the result. The set of constraints for `X` already contained a lower-bound constraint for `X` (`C <# X`), and that constraint has priority over the upper-bound "best-effort" approximation constraint, since type `C` is fully known (i.e. doesn't contain `_`).

With this feature, we derive a lower-bound constraint by combining the already existing lower-bound constraint with the actual bound.
```

I split the paragraph in two because this allows the last sentence to be more prominent, without the unusual ***doubly embellished font*** usage.

Author: Thanks for the suggestion, Erik! I've updated the text accordingly.

In general, if we have two type constraints of the form `E <: Y` and `Y <: F`, where `Y` is a type variable and `E` and `F` are type schemas, it follows that `E <: F` holds if the `E <: Y` and `Y <: F` hold. So, what if instead of adding the constraint `X <: B'` in step 5 of CSSTV, where `B'` is the best-effort bound approximation, we generate a new set of constraints, running the subtype constraint generation algorithm described in [Subtype constraint generation][subtype-constraint-generation] for `P = C`, `Q = B`, `L = {..., X, ...}`.

> `C` doesn't contain type variables from `L` by construction, and `B` may contain type variables from `L` (confidently, `X` in case X is F-bounded, but also any type variables from the same declaration, regardless of F-boundness of `X`), which makes `C <# B` a valid input for [Subtype constraint generation][subtype-constraint-generation] and a good source of additional type information.

Reviewer 2: ```suggestion
By construction, `C` doesn't contain type variables from `L`. `B` may contain type variables from `L` (surely, it contains `X` in case where `X` is F-bounded, but it may also contain other type variables from the same declaration, whether or not `X` is F-bounded), which makes `C <# B` a valid input for [Subtype constraint generation][subtype-constraint-generation] and a good source of additional type information.
```

Author: Thanks, Erik! I've updated the text as suggested.

The new constraints, generated by [Subtype constraint generation][subtype-constraint-generation] from `C <# B` are then added to the overall set of constraints for the current phase of type inference. Then, the solution for type variable `X` is computed from the new constraint set, at the end of the updated step 5 of CSSTV. 
Note that when `X` isn't F-bounded and its bound doesn't refer to other type variables from the same declaration, the suggested changes for step 5 of CSSTV work exactly as the unmodified step 5 of the same algorithm.

## Re-evaluating the motivating example

> Let's see how type inference, updated as proposed, works in the motivating example.

Reviewer 2: I think an empty line would be expected here:
```suggestion

Let's see how type inference, updated as proposed, works in the motivating example.
```

Author: Thanks for the suggestion! I've updated the file accordingly.

The picture below shows the details of the updated type inference process for the example.

```mermaid
block-beta
  columns 10
  space downwards["Downwards phase"]:4 upwards["Upwards phase"]:4 space

  space:5
  newConstraint["B <: X"]
  oldConstraint["C <: X"]
  space:3

  overallInput["Input"]
  space
  csstvDownwards["CSSTV"]
  space:4
  csstvUpwards["CSSTV"]
  space
  overallOutput["Output"]

  space:10

  space:5
  constraintGeneration["C <# A&lt;X&gt; (A&lt;B&gt; <# A&lt;X&gt;)"]
  space:4

  space:2
  boundsDownwards["X extends A<X>"]
  space:4
  boundsUpwards["X extends A <X>"]
  space:2

  downwards -- "∅" --> csstvDownwards
  upwards -- "C <: X" --> csstvUpwards

  overallInput -- "{X = _}" --> csstvDownwards
  csstvDownwards -- "{X = _}" --> csstvUpwards
  csstvUpwards -- "{X = B}" --> overallOutput

  oldConstraint --> constraintGeneration
  constraintGeneration --> newConstraint
  boundsUpwards --> constraintGeneration

  boundsDownwards --> csstvDownwards
  boundsUpwards --> csstvUpwards
```

## More examples enabled by the current proposal

> In addition to the motivating examples, both the original and the more complex, being unbroken by the current proposal, some cases of non F-bounded type variables are enabled by it as well.

Reviewer 2: ```suggestion
In addition to the motivating examples, both the original and the more complex, being handled by this feature, some cases of type variables that aren't F-bounded are handled as well.
```
Unbroken is a nice word, especially if it exists, but "being handled" is probably safer. ;-)

Author: Agreed! Thanks for the suggestion. I've updated the file accordingly.
